### PR TITLE
feat(stdlib): add encode_charset, decode_charset functions

### LINKED ISSUE #1155: feat(stdlib): Function for Converting Character Encodings
Our system utilizes both legacy systems and the latest Linux distributions. Occasionally, we encounter systems that operate with non-UTF-8 settings. When logs are sent via TCP from these systems, all Korean strings encoded in EUC-KR become corrupted.

To address this issue, I've created a functions.
The function signature is as follows.
```vrl
decode_charset(value, from_charset)
encode_charset(value, to_charset)
```

If this addition is beneficial for the VRL project, I would like to submit a pull request.

### PR DESCRIPTION
**Pull request by powerumc**
## Summary

Charset encoding and decoding functions were implemented to integrate the charset settings used in various operating systems and systems with vector.

- decode_charset(value, from_charset)
- encode_charset(value, to_charset)

## Change Type

- [ ] Bug fix
- [x] New feature
- [ ] Non-functional (chore, refactoring, docs)
- [ ] Performance

## Is this a breaking change?

- [ ] Yes
- [x] No

## How did you test this PR?

- [decode_charset unit test](https://github.com/powerumc/vector-vrl-contrib/blob/charset-encode-decode/src/stdlib/decode_charset.rs#L121-L155)
- [encode_charset unit test](https://github.com/powerumc/vector-vrl-contrib/blob/charset-encode-decode/src/stdlib/encode_charset.rs#L117-L151)

## Does this PR include user facing changes?

- [x] Yes. Please add a changelog fragment based on
  our [guidelines](https://github.com/vectordotdev/vector/blob/master/changelog.d/README.md).
- [ ] No. A maintainer will apply the "no-changelog" label to this PR.

## Checklist

- [x] Our [CONTRIBUTING.md](https://github.com/vectordotdev/vrl/blob/main/CONTRIBUTING.md) is a good starting place.
- [x] If this PR introduces changes to [LICENSE-3rdparty.csv](https://github.com/vectordotdev/vrl/blob/main/LICENSE-3rdparty.csv), please
  run `dd-rust-license-tool write` and commit the changes. More details [here](https://crates.io/crates/dd-rust-license-tool).
- [x] For new VRL functions, please also create a sibling PR in Vector to document the new function.
  - [vectordotdev/vector #21981](https://github.com/vectordotdev/vector/pull/21981) 

PR adding new VRL function: 
- [decode_charset](https://github.com/vectordotdev/vrl/pull/1162/files#diff-a2b29ab68373ed4a5433b9ccae091de7250ac1a3843df8700da2f382667b534c) function
- [encode_charset](https://github.com/vectordotdev/vrl/pull/1162/files#diff-e85c3cc34d764ef2b07194b6b6e6461b165064a3bda596d1c5124afa47aaf643) function

PR adding documentation:
- [1162.feature.md](https://github.com/vectordotdev/vrl/pull/1162/files#diff-98f42faae0e58a0579966173e58ea210c230b547ce8e6f5152c440422be07d76)

## References

- Closes: #1155

### CODE REVIEW

File: src/stdlib/decode_charset.rs [hunk 1]
```diff
@@ -0,0 +1,153 @@
...
+}
+
+fn decode_charset(value: &[u8], from_charset: &[u8]) -> Resolved {
+    let decoder = Encoding::for_label(from_charset).ok_or_else(|| error(from_charset))?;
+
+    let (output, _, _) = decoder.decode(value);
+    Ok(Value::Bytes(output.as_bytes().to_vec().into()))
+}
+
+fn error(from: &[u8]) -> ExpressionError {
```
Comment (pront) [inline src/stdlib/decode_charset.rs:87]: nit: 
```suggestion
fn create_error(from_charset: &[u8]) -> ExpressionError {
```

**Review by pront — APPROVED** [review]
Thank you @powerumc!

Left a couple of minor comments.

File: src/stdlib/encode_charset.rs [hunk 2]
```diff
@@ -0,0 +1,149 @@
...
+}
+
+fn encode_charset(value: &str, to_charset: &[u8]) -> Resolved {
+    let encoder = Encoding::for_label(to_charset).ok_or_else(|| error(to_charset))?;
+
+    let (output, _, _) = encoder.encode(value);
+    Ok(Value::Bytes(output.as_bytes().to_vec().into()))
+}
+
+fn error(from: &[u8]) -> ExpressionError {
```
Comment (pront) [inline src/stdlib/encode_charset.rs:83]: Nit: This is identical to the one above. Maybe we can move everything under a new directory `src/stdlib/charset`

File: src/stdlib/decode_charset.rs (hunk 1 above)
Comment (powerumc) [reply src/stdlib/decode_charset.rs]: hank you for your interest. 
I have modified it in the commit. d4be5d7a5712e4bf6c89104c35795f8f0e719cfa

**Comment by powerumc** [comment]
@pront 
I submitted a sibling PR in Vector to document the new function. [vectordotdev/vector #21981](https://github.com/vectordotdev/vector/pull/21981)
Thank you for your review. 😊

**Comment by pront** [comment]
The licenses check fails, please use `dd-rust-license-tool write && git commit -am "dd-rust-license-tool write"` and push to fix.

**Comment by powerumc** [comment]
> The licenses check fails, please use `dd-rust-license-tool write && git commit -am "dd-rust-license-tool write"` and push to fix.

In my case, there is no change in the `dd-rust-license-tool write` and `commit -am "dd-rust-license-tool write"`  command.
the `check-licenses` step in GitHub Actions was successful.

The error occurs in the check-deny step. Is this caused by my code?
https://github.com/vectordotdev/vrl/actions/runs/12160810717/job/34141068366?pr=1162

**Comment by jszwedko** [comment]
I think we just needed `main` merged in to fix the `cargo deny` failure. I merged it here.

**Pull request merged**